### PR TITLE
Add support for IPAdapterFull

### LINKED ISSUE #5886: [IP-Adapter] support face model
We recently added IP-adapter support to many of our pipelines in diffusers! You can now very easily load your IP-Adapter into a diffusers pipeline with `pipe.load_ip_adapter(...)`; 

However right now we do not support [`IPAdapterFull`](https://github.com/tencent-ailab/IP-Adapter/blob/e317f750e1829e846c944b9fdd7a2f2842c46501/ip_adapter/ip_adapter.py#L314) model checkpoints. Anyone interested to add that support for this? It will allow us to use the very cool face models with diffusers! 

the goal would be to make [this example](https://github.com/tencent-ailab/IP-Adapter/blob/main/ip_adapter-full-face_demo.ipynb) work with diffusers API. e.g.
```python
from diffusers import StableDiffusionPipeline
import torch
from diffusers.utils import load_image

base_model_path = "dreamlike-art/dreamlike-anime-1.0/"
pipeline = StableDiffusionPipeline.from_pretrained(
    base_model_path, torch_dtype=torch.float16)
pipeline.to("cuda")

image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ai_face2.png")

pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter-full-face_sd15.bin")
pipeline.set_ip_adapter_scale(0.6)
generator = torch.Generator(device="cpu").manual_seed(33)
images = pipeline(
    prompt="A photo of a girl wearing a black dress, holding red roses in hand, upper body, behind is the Eiffel Tower",
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50,
    generator=generator,
).images
images[0]
```

 You will need to look into [this PR](https://github.com/huggingface/diffusers/pull/5713/files#diff-272077719e5ba4349d4f0ee80bdc959ef954934af1df782deac4d297e359c08a) to understand how `load_ip_adapter` works and then make it work for [`IPAdapterFull`](https://github.com/tencent-ailab/IP-Adapter/blob/e317f750e1829e846c944b9fdd7a2f2842c46501/ip_adapter/ip_adapter.py#L314) model from the original repo.

### PR DESCRIPTION
**Pull request by fabiorigano**
# What does this PR do?

Fixes #5886

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@yiyixuxu

### CODE REVIEW

**Comment by fabiorigano** [comment]
@yiyixuxu sorry for disturbing, can I ask for feedback on this PR? 

I saw a newer PR is adding support for IP-Adapter Plus, should I add here too, as I only have to update checkpoint loading code (image embeddings are computed in the same way)?

Thanks

**Comment by fabiorigano** [comment]
@yiyixuxu perfect, I have removed IP-Adapter Plus from my PR

Example n. 1 -> SG161222/Realistic_Vision_V4.0_noVAE + ip-adapter full face + vae and ddim scheduler:

```from diffusers import StableDiffusionPipeline, AutoencoderKL, DDIMScheduler
from transformers import CLIPVisionModelWithProjection
import torch
from diffusers.utils import load_image
base_model_path = "SG161222/Realistic_Vision_V4.0_noVAE"
vae_model_path = "stabilityai/sd-vae-ft-mse"
image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", 
    subfolder="models/image_encoder",
    torch_dtype=torch.float16,
).to("cuda")
noise_scheduler = DDIMScheduler(
    num_train_timesteps=1000,
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    clip_sample=False,
    set_alpha_to_one=False,
    steps_offset=1,
)
vae = AutoencoderKL.from_pretrained(vae_model_path).to(dtype=torch.float16)
pipeline = StableDiffusionPipeline.from_pretrained(
    base_model_path,
    torch_dtype=torch.float16,
    image_encoder=image_encoder,
    feature_extractor=None,
    safety_checker=None
)
pipeline.to("cuda")
image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ai_face2.png")
pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter-full-face_sd15.bin")
pipeline.set_ip_adapter_scale(0.7)
generator = torch.Generator(device="cpu").manual_seed(33)
images = pipeline(
    prompt="A photo of a girl wearing a black dress, holding red roses in hand, upper body, behind is the Eiffel Tower",
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50, num_images_per_prompt=2,
    generator=generator,
).images
```

## Input
ai_face2

## Output
sg_1
sg_2

**Comment by fabiorigano** [comment]
Same initialization, different output width and height:
```images = pipeline(
    prompt="A photo of a girl wearing a black dress, holding red roses in hand, upper body, behind is the Eiffel Tower",
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50, num_images_per_prompt=2, width=512, height=704, 
    generator=generator,
).images
```


### Output

p0
p1

**Comment by fabiorigano** [comment]
Example n. 2 -> runwayml/stable-diffusion-v1-5 + ip-adapter full face + vae and ddim scheduler (width=512, height=704):

```
from diffusers import StableDiffusionPipeline, AutoencoderKL, DDIMScheduler
from transformers import CLIPVisionModelWithProjection
import torch
from diffusers.utils import load_image
base_model_path = "runwayml/stable-diffusion-v1-5"
vae_model_path = "stabilityai/sd-vae-ft-mse"
image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", 
    subfolder="models/image_encoder",
    torch_dtype=torch.float16,
).to("cuda")
noise_scheduler = DDIMScheduler(
    num_train_timesteps=1000,
    beta_start=0.00085,
    beta_end=0.012,
    beta_schedule="scaled_linear",
    clip_sample=False,
    set_alpha_to_one=False,
    steps_offset=1,
)
vae = AutoencoderKL.from_pretrained(vae_model_path).to(dtype=torch.float16)
pipeline = StableDiffusionPipeline.from_pretrained(
    base_model_path,
    torch_dtype=torch.float16,
	scheduler=noise_scheduler,
    vae=vae,
    image_encoder=image_encoder,
    feature_extractor=None,
    safety_checker=None
)
pipeline.to("cuda")
image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ai_face2.png")
pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter-full-face_sd15.bin")
pipeline.set_ip_adapter_scale(0.7)
generator = torch.Generator(device="cpu").manual_seed(33)
images = pipeline(
    prompt="A photo of a girl wearing a black dress, holding red roses in hand, upper body, behind is the Eiffel Tower",
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50, num_images_per_prompt=2, width=512, height=704,
    generator=generator,
).images
```

### Output
p0
p1

**Comment by fabiorigano** [comment]
After running python utils/check_copies.py --fix_and_overwrite, lines 80 and 81 in `src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py` are totally removed.

How can I solve it in order to pass all checks? @yiyixuxu

**Comment by fabiorigano** [comment]
> awesome! is the output in example n3 (with default scheduler and vae) expected?

@yiyixuxu yes, the original implementation with runwayml/stable-diffusion-v1-5, default ddim and vae gives the same outputs as example n 3. Using the other checkpoint (SG161222/Realistic_Vision_V4.0_noVAE), output images are "_clean_"

### Output from original IP-Adapter (runwayml/stable-diffusion-v1-5, default scheduler and vae)
image

**Comment by yiyixuxu** [comment]
can you add one example to use face model to the doc? https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters

and make it clear that we should not use default vae and scheduler for face models

cc @xiaohu2015 - why do we need to use different vae and scheduler for face models?

**Comment by xiaohu2015** [comment]
> can you add one example to use face model to the doc here https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters and make it clear that we should not use default vae and scheduler for face models
> 
> cc @xiaohu2015 here - why do we need to use different vae and scheduler for face models?

maybe vae is not so important, I think noise scheduler is more important?

**Comment by fabiorigano** [comment]
> can you add one example to use face model to the doc here https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters
> and make it clear that we should not use default vae and scheduler for face models

@yiyixuxu done!

how should I manage the output image? now the link points to one image in this PR, but maybe it should be moved here [huggingface/documentation-images](https://huggingface.co/datasets/huggingface/documentation-images). thanks

File: src/diffusers/models/embeddings.py [hunk 1]
```diff
@@ -460,6 +460,22 @@ def forward(self, image_embeds: torch.FloatTensor):
         return image_embeds
 
 
+class MLPProjection(nn.Module):
+    def __init__(self, image_embed_dim=1024, cross_attention_dim=1024):
+        super().__init__()
+
+        self.image_embeds = torch.nn.Sequential(
```
Comment (patrickvonplaten) [inline src/diffusers/models/embeddings.py:467]: Isn't this the same as https://github.com/huggingface/diffusers/blob/636feba552ab1f9a1ecdd3748e1245fa45b76d0d/src/diffusers/models/attention.py#L342

Could we try to re-use this module instead?

Comment (fabiorigano) [reply src/diffusers/models/embeddings.py:467]: hi @patrickvonplaten, I cannot use FeedForward because of a circular dependency between attention.py and embeddings.py, but I can write as follows to remove nn.Sequential

```
class MLPProjection(nn.Module):
    def __init__(self, image_embed_dim=1024, cross_attention_dim=1024):
        super().__init__()

        self.net = nn.ModuleList([
            nn.Linear(image_embed_dim, image_embed_dim),
            nn.GELU(),
            nn.Linear(image_embed_dim, cross_attention_dim),
            nn.LayerNorm(cross_attention_dim),
        ])

    def forward(self, image_embeds: torch.FloatTensor):
        for module in self.net:
            image_embeds = module(image_embeds)
        return image_embeds
```

Comment (fabiorigano) [reply src/diffusers/models/embeddings.py:467]: hi again, I have found a solution, you were right @patrickvonplaten 

```
class MLPProjection(nn.Module):
    def __init__(self, image_embed_dim=1024, cross_attention_dim=1024):
        super().__init__()
        from .attention import FeedForward
        
        self.ff = FeedForward(image_embed_dim, cross_attention_dim, mult=1, activation_fn="gelu")
        self.norm = nn.LayerNorm(cross_attention_dim)

    def forward(self, image_embeds: torch.FloatTensor):
        return self.norm(self.ff(image_embeds))
```

File: docs/source/en/using-diffusers/loading_adapters.md [hunk 2]
```diff
@@ -485,6 +485,69 @@ image.save("sdxl_t2i.png")
...
+    set_alpha_to_one=False,
+    steps_offset=1
+)
+
+pipeline = StableDiffusionPipeline.from_pretrained(
+    "runwayml/stable-diffusion-v1-5",
+    torch_dtype=torch.float16,
+    scheduler=noise_scheduler,
+    feature_extractor=None,
+    safety_checker=None
```
Comment (yiyixuxu) [inline docs/source/en/using-diffusers/loading_adapters.md:521]:
```suggestion
```

**Review by yiyixuxu — COMMENTED** [review]
thanks for working on this! 
I left a few comments, overall looking great and can't wait to have this in diffusers

might have to coordinate with https://github.com/huggingface/diffusers/pull/5915#issuecomment-1831290766 in terms of merging and may need to rebase

**Comment by fabiorigano** [comment]
where can I upload output image for documentation?

**Comment by fabiorigano** [comment]
> thanks for working on this! I left a few comments, overall looking great and can't wait to have this in diffusers
> 
> might have to coordinate with #5915 in terms of merging and may need to rebase

How do you suggest we proceed with merge? thanks

**Comment by yiyixuxu** [comment]
> where can I upload output image for documentation?

cc @sayakpaul how is it normally done? do contributors send a PR to our repo that hosts documentation images?

**Comment by sayakpaul** [comment]
Yes

**Comment by yiyixuxu** [comment]
@fabiorigano 
for documentation image, you can send a PR here https://huggingface.co/datasets/huggingface/documentation-images/tree/main/diffusers

**Comment by fabiorigano** [comment]
> @fabiorigano for documentation image, you can send a PR here https://huggingface.co/datasets/huggingface/documentation-images/tree/main/diffusers

@yiyixuxu done, https://huggingface.co/datasets/huggingface/documentation-images/discussions/234

I accidentally opened another PR in the wrong directory, I can only close it tomorrow due to restrictions on new users

**Comment by yiyixuxu** [comment]
@sayakpaul can we merge this?

> @yiyixuxu done, https://huggingface.co/datasets/huggingface/documentation-images/discussions/234

**Review by yiyixuxu — APPROVED** [review]
looks good to merge to me :)

**Comment by yiyixuxu** [comment]
@patrickvonplaten 
can we have a final review here? this is the last piece of ip-adapter!

**Comment by fabiorigano** [comment]
> will merge now but need to open a follow up PR for:
> 
> 1. [[Feature] Support IP-Adapter Plus #5915 (comment)](https://github.com/huggingface/diffusers/pull/5915#discussion_r1413752931)
> 2. [Add support for IPAdapterFull #5911 (comment)](https://github.com/huggingface/diffusers/pull/5911#discussion_r1418063329)

@yiyixuxu I can open a new PR if it is ok then

**Comment by fabiorigano** [comment]
@yiyixuxu @patrickvonplaten thank you so much for merging my first-ever pull request! I appreciate your time and effort in reviewing and guiding the changes :) 
I look forward to continuing to contribute to this community
